### ISSUE TITLE
Add narrative documentation showing how to use twisted.names to create a custom server

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6864]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6864|
|Type|enhancement|
|Created|2013-12-07 13:40:47Z|
|Branch|https://github.com/twisted/twisted/tree/custom-server-documentation-6864-2|

The twisted.names narrative documentation shows how to load an authoritative server from a pyzone file and twistd.

But it doesn't demonstrate how t.names.authority can be used to create a custom authoritative server.

Nor does it show how authoritative server can be combined with other resolvers and how certain results can be overridden.

hynek showed me (rwall) some interesting example code at the Bristol sprint which would be a good starting point for this ticket.

Attachments:

* [auth-dns-docs-6864.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9bb/9bb2b703661ba5c7a37b889f18822e146cca985c/8907a1cc9c98da47c7ec1f008a1beb3685d5298e.patch) (11525 bytes) - added by rwall on 2014-01-14 10:28:16Z - An example of a custom DNS resolver with pass through

<details><summary>Searchable metadata</summary>

```
trac-id__6864 6864
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_custom_server_documentation_6864_2 branches/custom-server-documentation-6864-2
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__documentation documentation
time__1386423647000000 1386423647000000
changetime__1390036154000000 1390036154000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

(In [41236]) Branching to 'custom-server-documentation-6864'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

(In [41272]) Branching to 'custom-server-documentation-6864-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in log:branches/custom-server-documentation-6864-2
 * Added some examples of custom (non-authoritative) DNS servers
 * And some brief narrative documentation explaining how they work

Over to you hynek :)

If you can describe the other case that you showed me in Bristol I might try and describe that too....here or in another ticket.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn set owner to @khorn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_6"></a>|@khorn set owner to @wallrj|
|-|-|

Review Notes:

 * I don't see simple_server.py or override_server.py.  Did you forget to check them in, or am I just blind?
 * Note that in RestructuredText when marking up a literal block (using ```::```), you can put the ```::``` at the end of the preceding line, similar to how one might use a colon to lead into the block.  So instead of:
```
here's some code
::

    code goes here
```
you'd have something like:
```
here's some code::

    code goes here
```

I think this makes the markup look a bit nicer, and would recommend changing to this style, but don't block the ticket on this if you disagree.
 * Maybe I'm being a bit overly-pedantic, but in yoru explanatory note section, you say:
```
It takes a list of :api:`twisted.internet.interfaces.IResolver <IResolver>` providers and queries each one in turn until it receives an answer.
```
Perhaps it should be more exact, like:
```
It takes a list of :api:`twisted.internet.interfaces.IResolver <IResolver>` providers and queries each one in turn until either it receives an answer, or they all fail.
```
Probably you can make that sentence flow  a little better, but I think it should be made clear that it isn't going to repeat the chain over and over.  Maybe this change isn't strictly necessary, but I think it makes things a little more clear for those who aren't very familiar with Twisted.

The narrative itself looks pretty good, but it's difficult to say without seeing the actual code examples. Please add the code or point out what I'm missing and resubmit for review.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn set owner to @khorn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

Thanks for the review Kevin.

My comments and answers are below...

Replying to [khorn](#note_6):
> Review Notes:
>  * I don't see simple_server.py or override_server.py.  Did you forget to check them in, or am I just blind?

They are there, but may not be shown in the Trac diff because they're new files.
Trac does show the links to those files at the top of the diff page though (in green).

>  * Note that in RestructuredText when marking up a literal block (using ```::```), you can put the ```::``` at the end of the preceding line, similar to how one might use a colon to lead into the block.  So instead of:

Yeah, I had tried that, but I was confused because it renedered as a single : at
the end of the line, or sometimes a :: if there was a preceding full stop.

And neither form resulted in syntax highlighting, so I've switched to the
'.. code-block:: console' syntax instead.

> Probably you can make that sentence flow  a little better, but I think it should be made clear that it isn't going to repeat the chain over and over.  Maybe this change isn't strictly necessary, but I think it makes things a little more clear for those who aren't very familiar with Twisted.

I agree. I've changed it. And added a little more detail about the use of DomainError.

Ready for another review.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_10"></a>|@khorn set owner to @wallrj|
|-|-|

This looks great and could be merged as it is, but here's a couple of things which might improve it:

 * Some of the lines in the RestructuredText files are pretty long, and should probably be shortened
 * In the `override_server.py` example, in `DynamicResolver._doDynamicResponse` the return values are a bit opaque.  I'm not too concerned about this, since it is explained in the narrative text, but you might consider adding each of the three return values to temporary/throwaway variables with appropriate names, and then returning those names, just to help clarify things.

Other than that, looks great.  Merge when ready.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_11"></a>|@wallrj set status to `closed`|
|-|-|

(In [41335]) Merge custom-server-documentation-6864-2

Author: rwall
Reviewer: khorn
Fixes: [#6864](https://github.com/twisted/twisted/issues/6864)

Added some narrative documentation showing how to create a custom DNS
server. Also added a few introductory paragraphs to the twisted.names
documentation index page.
